testscript/exe.go

```diff
@@ -49,63 +51,138 @@ func IgnoreMissedCoverage() {
+			if runtime.GOOS == "windows" {
+				binfile += ".exe"
+			}
+			binpath, err := os.Executable()
+			if err == nil {
+				err = copyBinary(binpath, binfile)
+			}
+			if err != nil {
+				log.Printf("could not set up %s in $PATH: %v", name, err)
+				return 2
+			}
 			scriptCmds[name] = func(ts *TestScript, neg bool, args []string) {
-				path, err := os.Executable()
-				if err != nil {
-					ts.Fatalf("cannot determine path to test binary: %v", err)
-				}
-				id := atomic.AddInt32(&profileId, 1) - 1
-				oldEnvLen := len(ts.env)
-				cprof := coverFilename(id)
-				ts.env = append(ts.env,
-					"TESTSCRIPT_COMMAND="+name,
-					"TESTSCRIPT_COVERPROFILE="+cprof,
-				)
-				ts.cmdExec(neg, append([]string{path}, args...))
-				ts.env = ts.env[0:oldEnvLen]
-				if cprof == "" {
-					return
-				}
-				f, err := os.Open(cprof)
-				if err != nil {
-					if ignoreMissedCoverage {
-						return
-					}
-					ts.Fatalf("command %s (args %q) failed to generate coverage information", name, args)
-					return
-				}
-				coverChan <- f
+				ts.cmdExec(neg, append([]string{name}, args...))
 			}
 		}
 		return m.Run()
 	}
-	mainf := commands[cmdName]
-	if mainf == nil {
-		log.Printf("unknown command name %q", cmdName)
-		return 2
-	}
 	// The command being registered is being invoked, so run it, then exit.
 	os.Args[0] = cmdName
-	cprof := os.Getenv("TESTSCRIPT_COVERPROFILE")
-	if cprof == "" {
+	coverdir := os.Getenv("TESTSCRIPT_COVER_DIR")
+	if coverdir == "" {
 		// No coverage, act as normal.
 		return mainf()
 	}
+
+	// For a command "foo", write ${TESTSCRIPT_COVER_DIR}/foo-${RANDOM}.
+	// Note that we do not use ioutil.TempFile as that creates the file.
+	// In this case, we want to leave it to -test.coverprofile to create the
+	// file, as otherwise we could end up with an empty file.
+	// Later, when merging profiles, trying to merge an empty file would
+	// result in a confusing error.
+	rnd, err := nextRandom()
+	if err != nil {
+		log.Printf("could not obtain random number: %v", err)
+		return 2
+	}
+	cprof := filepath.Join(coverdir, fmt.Sprintf("%s-%x", cmdName, rnd))
 	return runCoverSubcommand(cprof, mainf)
 }
 
+func nextRandom() ([]byte, error) {
+	p := make([]byte, 4)
+	_, err := cryptorand.Read(p)
+	return p, err
+}
+
+// copyBinary makes a copy of a binary to a new location. It is used as part of
+// setting up top-level commands in $PATH.
+//
+// It does not attempt to use symlinks for two reasons:
+//
+// First, some tools like cmd/go's -toolexec will be clever enough to realise
+// when they're given a symlink, and they will use the symlink target for
+// executing the program. This breaks testscript, as we depend on os.Args[0] to
+// know what command to run.
+//
+// Second, symlinks might not be available on some environments, so we have to
+// implement a "full copy" fallback anyway.
+func copyBinary(from, to string) error {
+	writer, err := os.OpenFile(to, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0777)
```

Comment:
tbh if you need O_TRUNC here, something is going wrong...

Reply:
That was just a copy-paste from os.Create :) I don't think it matters either way, so I'll remove it.